autogen/coding/base.py

```diff
@@ -0,0 +1,94 @@
+from typing import Any, Dict, List, Protocol, Union, runtime_checkable
+
+from pydantic import BaseModel, Field
+
+from ..agentchat.agent import LLMAgent
+
+__all__ = ("CodeBlock", "CodeResult", "CodeExtractor", "CodeExecutor")
+
+
+class CodeBlock(BaseModel):
+    """A class that represents a code block."""
+
+    code: str = Field(description="The code to execute.")
+
+    language: str = Field(description="The language of the code.")
+
+
+class CodeResult(BaseModel):
+    """A class that represents the result of a code execution."""
+
+    exit_code: int = Field(description="The exit code of the code execution.")
+
+    output: str = Field(description="The output of the code execution.")
+
+
+class CodeExtractor(Protocol):
+    """A code extractor class that extracts code blocks from a message."""
+
+    def extract_code_blocks(self, message: Union[str, List[Dict[str, Any]], None]) -> List[CodeBlock]:
+        """Extract code blocks from a message.
+
+        Args:
+            message (str): The message to extract code blocks from.
+
+        Returns:
+            List[CodeBlock]: The extracted code blocks.
+        """
+        ...  # pragma: no cover
+
+
+@runtime_checkable
+class CodeExecutor(Protocol):
+    """A code executor class that executes code blocks and returns the result."""
+
+    class UserCapability(Protocol):
+        """An AgentCapability class that gives agent ability use this code executor."""
+
+        def add_to_agent(self, agent: LLMAgent) -> None:
+            ...  # pragma: no cover
+
+    @property
+    def user_capability(self) -> "CodeExecutor.UserCapability":
+        """Capability to use this code executor.
+
+        The exported capability can be added to an agent to allow it to use this
+        code executor:
+
+        ```python
+        code_executor = CodeExecutor()
+        agent = ConversableAgent("agent", ...)
+        code_executor.user_capability.add_to_agent(agent)
+        ```
+
+        A typical implementation is to update the system message of the agent with
+        instructions for how to use this code executor.
+        """
+        ...  # pragma: no cover
+
+    @property
+    def code_extractor(self) -> CodeExtractor:
+        """The code extractor used by this code executor."""
+        ...  # pragma: no cover
+
+    def execute_code_blocks(self, code_blocks: List[CodeBlock]) -> CodeResult:
```

Comment:
I'm exploring the design intentions behind the execute_code_blocks method, particularly in regards to its apparent focus on supporting interpreted languages. Given that this method facilitates the execution of multiple code blocks with variable sharing and state persistence across these blocks, does this design inherently favor interpreted languages like Python, JavaScript, etc.? Furthermore, I'm curious about whether this design choice limits the range of supported programming languages, especially considering the distinct execution models of compiled languages (e.g., C, C++, Go, Rust) compared to interpreted ones. How does this design accommodate or adapt to the needs of compiled languages in maintaining a coherent state across code blocks?

Reply:
For compiled language we would have to use a specific instruction to instruct the LLM to write each code block as a file. The code executor then compiles the files in `execute_code_blocks` and run them. 

Many compiled languages have interactive version (e.g., .NET Interactive). I think under conversational setting it is easier to use languages that support interactive coding and LLM is probably better tuned for this type of coding as well.

Reply:
OK. Supporting compiled languages will be more complicated. For example, you need to design a mechanism that pays attention to variable transfer, which will be a relatively large task.